File: core/data/src/main/java/org/mifos/mobilewallet/core/domain/usecase/user/CreateUser.java
package org.mifos.mobilewallet.core.domain.usecase.user;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.domain.model.user.NewUser;
import org.mifos.mobilewallet.core.utils.ErrorJsonMessageHelper;

import javax.inject.Inject;

import retrofit2.HttpException;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by naman on 17/6/17.
 */

public class CreateUser extends UseCase<CreateUser.RequestValues, CreateUser.ResponseValue> {

    private final FineractRepository apiRepository;

    @Inject
    public CreateUser(FineractRepository apiRepository) {
        this.apiRepository = apiRepository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {
        apiRepository.createUser(requestValues.user)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<CreateUser.ResponseValue>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        ErrorJsonMessageHelper.getUserMessage(e);
                        String message = "Error";
                        try {
                            message = ((HttpException) e).response().errorBody().string();
                            message = ErrorJsonMessageHelper.getUserMessage(message);
                        } catch (Exception e1) {
                            message = "Error";
                        }
                        getUseCaseCallback().onError(message);
                    }

                    @Override
                    public void onNext(CreateUser.ResponseValue genericResponse) {
                        getUseCaseCallback().onSuccess(genericResponse);
                    }
                });
    }

    public static final class RequestValues implements UseCase.RequestValues {

        private final NewUser user;

        public RequestValues(NewUser user) {
            this.user = user;
        }

        public NewUser getUser() {
            return user;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        private final int resourceId;

        public ResponseValue(int userId) {
            this.resourceId = userId;
        }

        public int getUserId() {
            return resourceId;
        }
    }

}

File: core/data/src/main/java/org/mifos/mobilewallet/core/domain/usecase/user/CreateUser.kt
package org.mifos.mobilewallet.core.domain.usecase.user

import org.mifos.mobilewallet.core.base.UseCase
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository
import org.mifos.mobilewallet.core.domain.model.user.NewUser
import org.mifos.mobilewallet.core.utils.ErrorJsonMessageHelper.getUserMessage
import retrofit2.HttpException
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by naman on 17/6/17.
 */
class CreateUser @Inject constructor(private val apiRepository: FineractRepository) :
    UseCase<CreateUser.RequestValues, CreateUser.ResponseValue>() {

    override fun executeUseCase(requestValues: RequestValues) {
        apiRepository.createUser(requestValues.user)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<ResponseValue>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    getUserMessage(e)
                    var message: String? = "Error"
                    try {
                        message = (e as HttpException).response().errorBody().string()
                        message = getUserMessage(message)
                    } catch (e1: Exception) {
                        message = "Error"
                    }
                    useCaseCallback.onError(message)
                }

                override fun onNext(genericResponse: ResponseValue) {
                    useCaseCallback.onSuccess(genericResponse)
                }
            })
    }

    class RequestValues(val user: NewUser) : UseCase.RequestValues
    class ResponseValue(val userId: Int) : UseCase.ResponseValue
}
